From 5735f0d989b20a772de9a0e93536ee2387686560 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:53 -0400
Subject: [usb] xhci: Set SCT field for Set TR dequeue on streams

Message-id: <1410887463-20674-282-git-send-email-dzickus@redhat.com>
Patchwork-id: 95217
O-Subject: [RHEL7 PATCH 281/591] xhci: Set SCT field for Set TR dequeue on streams
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 95241dbdf8281ece1355b8673b882d6a182f3c7d
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Fri Oct 4 00:29:48 2013 +0200

    xhci: Set SCT field for Set TR dequeue on streams

    Nec XHCI controllers don't seem to care, but without this Intel XHCI
    controllers reject Set TR dequeue commands with a COMP_TRB_ERR, leading
    to the following warning:

    WARN Set TR Deq Ptr cmd invalid because of stream ID configuration

    And very shortly after this the system completely freezes.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-ring.c b/drivers/usb/host/xhci-ring.c
index 7c2bbe3..7820c76 100644
--- a/drivers/usb/host/xhci-ring.c
+++ b/drivers/usb/host/xhci-ring.c
@@ -4071,6 +4071,7 @@ static int queue_set_tr_deq(struct xhci_hcd *xhci, int slot_id,
  u32 trb_slot_id = SLOT_ID_FOR_TRB(slot_id);
  u32 trb_ep_index = EP_ID_FOR_TRB(ep_index);
  u32 trb_stream_id = STREAM_ID_FOR_TRB(stream_id);
+ u32 trb_sct = 0;
  u32 type = TRB_TYPE(TRB_SET_DEQ);
  struct xhci_virt_ep *ep;
 
@@ -4089,7 +4090,9 @@ static int queue_set_tr_deq(struct xhci_hcd *xhci, int slot_id,
  }
  ep->queued_deq_seg = deq_seg;
  ep->queued_deq_ptr = deq_ptr;
- return queue_command(xhci, lower_32_bits(addr) | cycle_state,
+ if (stream_id)
+  trb_sct = SCT_FOR_TRB(SCT_PRI_TR);
+ return queue_command(xhci, lower_32_bits(addr) | trb_sct | cycle_state,
    upper_32_bits(addr), trb_stream_id,
    trb_slot_id | trb_ep_index | type, false);
 }
diff --git a/drivers/usb/host/xhci.h b/drivers/usb/host/xhci.h
index 152adba..c7619d7 100644
--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -1129,9 +1129,10 @@ enum xhci_setup_dev {
 #define TRB_TO_SUSPEND_PORT(p)  (((p) & (1 << 23)) >> 23)
 #define LAST_EP_INDEX   30
 
-/* Set TR Dequeue Pointer command TRB fields */
+/* Set TR Dequeue Pointer command TRB fields, 6.4.3.9 */
 #define TRB_TO_STREAM_ID(p)  ((((p) & (0xffff << 16)) >> 16))
 #define STREAM_ID_FOR_TRB(p)  ((((p)) & 0xffff) << 16)
+#define SCT_FOR_TRB(p)   (((p) << 1) & 0x7)
 
 
 /* Port Status Change Event TRB fields */
-- 
1.7.1